sdk/core/core-http/src/fetchHttpClient.ts

```diff
@@ -237,9 +240,12 @@ function isReadableStream(body: any): body is Readable {
   return body && typeof body.pipe === "function";
 }
 
-function isStreamComplete(stream: Readable): Promise<void> {
+function isStreamComplete(stream: Readable, aborter: AbortController): Promise<void> {
   return new Promise((resolve) => {
-    stream.on("close", resolve);
+    stream.on("close", () => {
```

Comment:
Is it possible to write some tests around this? It makes sense to me that if the download stream completes, we've already gotten the response so `aborting` is ok, but what happens when the uploadStream is closed... will this cause an `AbortError` to be thrown? Is the uploadSteam closed once the ClientRequest is done reading the data, or once the ClientRequest is closed as well?

Reply:
> but what happens when the uploadStream is closed... will this cause an `AbortError` to be thrown?

Yes it would cause an `AbortError`, which is one of my concern. For example, consumers call `stream.destroy()` and expect TCP connection to be closed. We don't seem to have other way to terminate the network connection?

More realistic scenario in the original issue: front end streaming an video from blob storage, when navigating away from the page, the backend should close the TCP connection to Azure Storage server.

Reply:
I'm worried about the scenario where the user uploads a file to storage blob. Once the ClientRequest is done reading from the file readable stream, does that close the file stream too then? I'm worried this will happen before we've gotten a response from the service, causing an AbortError to be thrown when we should get a successful response back (or an error from the service)

Reply:
Interesting point! I hadn't considered that.  You are talking about the scenario where
1. customer creates a readable stream from a file
2. customer passes the stream to `upload()`, which eventually was passed to our `request.body`
3. when file stream is read to the end, potentially it would be closed, thus causing the uploading request to be aborted with changes in this PR.

I will give it a try.  Another thing we could do is scope upload out of this PR until we have better understanding of upload scenarios.

Reply:
I changed to only abort when download stream is closed.

Reply:
Small nit that I think we should be using `once` instead of `on` here

Reply:
Updated

Reply:
> Is it possible to write some tests around this?

I tried to test using `PassThrough` as my stream but it doesn't have `destroy()` or `close()`. I ended up creating a local http server for the test.

@chradek @xirzec please have another look

Reply:
Thanks for the tests! If you wanted to use PassThrough, could you perhaps use a subclass of it to fill in the extra methods?

```ts
export class FakeHttp extends PassThrough {
  public closed?: boolean;
  public destroyed?: boolean;
  public destroy(): void {
    this.destroyed = true;
  }
  public close(): void {
    this.closed = true;
  }
}
```

Reply:
> If you wanted to use PassThrough, could you perhaps use a subclass of it to fill in the extra methods?

I gave it a try using above `FakeHttp` and our httpMock but it didn't work. It looks that node-fetch would convert the body into `Buffer`
https://github.com/node-fetch/node-fetch/blob/v2.6.5/src/body.js#L55

Also our mock converts stream to buffer even if a stream were able to reach there
https://github.com/Azure/azure-sdk-for-js/blob/584693d6533c7f0f71d07ddd7d1d8c69e793e2f8/sdk/core/core-http/test/mockHttp.ts#L96